sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -269,6 +285,14 @@ trait HiveTypeCoercion {
         i.makeCopy(Array(Cast(a, StringType), b.map(Cast(_, StringType))))
       case i @ In(a, b) if a.dataType == TimestampType && b.forall(_.dataType == DateType) =>
         i.makeCopy(Array(Cast(a, StringType), b.map(Cast(_, StringType))))
+      case i @ In(a, b) if a.dataType == StringType
+        && b.exists(_.dataType.isInstanceOf[NumericType]) =>
+        i.makeCopy(Array(Cast(a, DoubleType), b))
```

Comment:
As I've commented on the JIRA ticket, this is not the behavior of Hive. Hive actually converts the numerics in the constant set into strings.